cachito/web/models.py

```diff
@@ -346,6 +347,65 @@ class RequestDependency(db.Model):
     __table_args__ = (db.UniqueConstraint("request_id", "dependency_id", "package_id"),)
 
 
+def _validate_request_package_configs(request_kwargs, pkg_managers_names):
+    """
+    Validate the "packages" parameter in a new request.
+
+    :param dict request_kwargs: the JSON parameters of the new request
+    :param list pkg_managers_names: the list of valid package manager names for the request
+    :raises ValidationError: if the "packages" parameter is invalid
+    """
+    # Validate the custom packages configuration. For example:
+    # {"packages": {"npm": [{"path": "client"}]}}
+    packages_configs = request_kwargs.get("packages", {})
+    if not isinstance(packages_configs, dict):
+        raise ValidationError('The "packages" parameter must be an object')
+
+    invalid_package_managers = packages_configs.keys() - set(pkg_managers_names)
+    if invalid_package_managers:
+        raise ValidationError(
+            'The following package managers in the "packages" value do not apply: '
+            + ", ".join(invalid_package_managers)
+        )
+
+    supported_packages_configs = {"npm"}
+    unsupported_packages_managers = packages_configs.keys() - supported_packages_configs
+    if unsupported_packages_managers:
+        raise ValidationError(
+            'The following package managers in the "packages" value are unsupported: '
+            + ", ".join(unsupported_packages_managers)
+        )
+
+    # Validate the values for each package manager configuration (e.g. packages.npm)
+    valid_package_config_keys = {"path"}
+    for pkg_manager, packages_config in packages_configs.items():
+        invalid_format_error = (
+            f'The "packages.{pkg_manager}" value must be an array of objects with the following '
+            f'keys: {", ".join(valid_package_config_keys)}'
+        )
+        if not isinstance(packages_config, list):
+            raise ValidationError(invalid_format_error)
+
+        for package_config in packages_config:
+            if not isinstance(package_config, dict) or not package_config:
+                raise ValidationError(invalid_format_error)
+
+            invalid_keys = package_config.keys() - valid_package_config_keys
+            if invalid_keys:
+                raise ValidationError(invalid_format_error)
+
+            if not (
+                isinstance(package_config["path"], str)
+                and package_config["path"]
+                and not os.path.isabs(package_config["path"])
+                and os.path.pardir not in package_config["path"]
+            ):
+                raise ValidationError(
+                    f'The "path" values in the "packages.{pkg_manager}" value must be to a '
```

Comment:
This message may be confusing to users. A path with `..` in it could be resolved to a relative path withing the application repo, e.g. `foo/bar/../baz`. This would be unusual, but the error message wouldn't necessarily point the user in the right direction.

Reply:
Okay, I'll check it against `os.path.normpath(package_config["path"])`.